packages/editor/src/store/middlewares/sync-blocks-to-content.js

```diff
@@ -0,0 +1,107 @@
+/**
+ * External dependencies
+ */
+import { merge } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { parse } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import { resetBlocks } from '../actions';
+import { getPostEdits, getCurrentPostAttribute } from '../selectors';
+
+/**
+ * Returns true if the action includes the `skipContentParse` option indicating
+ * that content should be synced to blocks state, regardless of whether the
+ * value has changed.
+ *
+ * @param {Object} action Action object to test.
+ *
+ * @return {boolean} Whether block parse is to be forced.
+ */
+export function isForcedParse( action ) {
+	return Boolean( action.options ) && action.options.skipContentParse === false;
+}
+
+/**
+ * Returns true if the action includes the `skipContentParse` option indicating
+ * that a parse to sync blocks state should be skipped, even if content has
+ * changed. This can be used to optimize frequent updates to content, followed
+ * by a subsequent force parse.
+ *
+ * @param {Object} action Action object to test.
+ *
+ * @return {boolean} Whether block parse is to be forced.
+ */
+export function isSkippedParse( action ) {
+	return Boolean( action.options && action.options.skipContentParse );
+}
+
+/**
+ * Returns the current edited post content. Unlike the `getEditedPostContent`
+ * selector from the editor store, this intentionally disregards the blocks
+ * state, since it's assumed this would be used by the middleware in populating
+ * said blocks state.
+ *
+ * In all other usage of edited post content, the blocks state is considered
+ * the canonical source of truth. The middleware is responsible for making this
+ * blocks state available. For example, when editing an existing post, blocks
+ * state will have a default value of an empty array. In resetting the blocks
+ * state, the middleware will consider initial edits, if any exist, but defer
+ * to the post's persisted value.
+ *
+ * @param {Object} state Editor state.
+ *
+ * @return {string} Edited post content.
+ */
+export function getPostContent( state ) {
+	const edits = getPostEdits( state );
+	if ( edits.hasOwnProperty( 'content' ) ) {
+		return edits.content;
+	}
+
+	return getCurrentPostAttribute( state, 'content' );
+}
+
+export default ( store ) => {
+	const { getState } = store;
+	let previousContent;
+
+	return ( next ) => ( action ) => {
+		const result = next( action );
```

Comment:
Why do we need this to be a middleware? You probably guessed that I don't like middlewares too much :P because I think they obfuscate code even more than effects :). Also, it's one other pattern that may be replaced by something else.

If I understand properly it's a middleware for two reasons (correct me if I'm wrong): 

 - Because it's not really triggered by a single action but it's more triggered by a change in the state
 - Because it needs to perform side effects (async stuff)

Did you consider:

 - Making it a higher-order reducer. (I guess the argument against this is the upcoming async parser, instead of a required sync parser in reducers)
 - Making it an effect? How many actions need to trigger this?

Reply:
Can't this be a combination of an effect (adding the parsed content to the required actions) and a higher-order reducer using the parsed content?

Reply:
Yes, my original intention was that this didn't fall neatly into our other patterns (effects, generators) because it doesn't care which action triggered the change, it only cares about the change in the content state values. In retrospect, I'd probably agree that this is the exact sort of unpredictable behavior we might want to discourage.

An original thought I'd had was to support some kind of action creator to set content using our new routines generators (enabling both our current sync parse and a future async parse). Roughly:

```js
function* setContent( content ) {
	const blocks = yield { type: 'PARSE', content };
	yield resetBlocks( blocks );
}
```

In thinking more on it, it seemed a bit redundant considering we have other actions which operate on _any_ edit to a post property; having a separate one for "content" seemed out of place.

Should those action creators do the parse when the payload includes `content` ? What are those action creators? `editPost`, `updatePost`, `resetPost` ... ?

Worth thinking on more...